pennylane/dla/lie_closure.py

```diff
@@ -24,6 +25,61 @@
 from pennylane.pauli import PauliSentence, PauliWord
 
 
+def lie_closure(
+    generators: Iterable[Union[PauliWord, PauliSentence, Operator]],
+    max_iterations: int = 1e4,
+    verbose: int = 0,
+) -> Iterable[Union[PauliWord, PauliSentence, Operator]]:
+    r"""Compute the Lie closure over commutation of a set of generators
+
+    The Lie closure of a set of generators :math:`\mathcal{G} = \{G_1, .. , G_N\}` is computed by
+    taking all possible nested commutators until no new (i.e. linearly independent) operator is produced.
+
+    Args:
+        generators (Iterable[Union[PauliWord, PauliSentence, Operator]]): generating set for which to compute the
+            Lie closure.
+        max_iterations (int): maximum depth of nested commutators to consider.
+
+    Returns:
+        list[`~.PauliSentence`]: a basis of ``PauliSentence`` instances that is closed under
+        commutators (Lie closure).
+    """
+    if not all(isinstance(op, PauliSentence) for op in generators):
+        generators = [
+            op.pauli_rep if op.pauli_rep else qml.pauli.pauli_sentence(op) for op in generators
+        ]
+
+    vspace = PauliVSpace(generators)
+
+    epoch = 0
+    old_length = 0  # dummy value
+    new_length = len(vspace)
+
+    while new_length > old_length or epoch > max_iterations:
+        if verbose > 0:
+            print(f"epoch {epoch+1} of lie_closure")
+        for idx1 in range(new_length - 1):
+            ps1 = vspace.basis[idx1]
+            for idx2 in range(max([idx1 + 1, old_length]), new_length):
+                ps2 = vspace.basis[idx2]
```

Comment:
Nice! Do you know if this tends to be also fast (on top of being much more readable)?

Reply:
I guess in certain cases it could be faster, but I think the bottleneck here will be computing the commutators and augmenting the subspace basis in any case.